src/server/send.rs

```diff
@@ -130,32 +134,64 @@ pub fn send_dir<P1: AsRef<Path>, P2: AsRef<Path>>(
     // Sort files (dir-first and lexicographic ordering).
     files.sort_unstable();
 
-    Ok(render(dir_path.filename_str(), &files, &breadcrumbs).into())
+    let content = render(dir_path.filename_str(), &files, &breadcrumbs).into_bytes();
+    let size = content.len();
+    Ok((content, size))
 }
 
-/// Send a buffer of file to client.
-pub fn send_file<P: AsRef<Path>>(file_path: P) -> io::Result<Vec<u8>> {
-    use std::io::prelude::*;
-    let f = File::open(file_path)?;
-    let mut buffer = Vec::new();
-    BufReader::new(f).read_to_end(&mut buffer)?;
+const BUFFER_SIZE: usize = 4_096;
 
-    Ok(buffer)
+#[derive(Debug)]
+pub struct FileStream<T> {
+    reader: Mutex<T>,
+}
+
+impl<T: Read> Stream for FileStream<T> {
+    type Item = io::Result<hyper::body::Bytes>;
+
+    fn poll_next(self: Pin<&mut Self>, _: &mut std::task::Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut r = match self.reader.lock() {
+            Ok(r) => r,
+            Err(e) => {
+                eprintln!("{e:?}");
+                let e = io::Error::new(io::ErrorKind::Other, "Failed to read file");
+                return Poll::Ready(Some(Err(e)));
+            }
+        };
+        let mut buf: [u8; BUFFER_SIZE] = [0; BUFFER_SIZE];
+        match r.read(&mut buf) {
+            Ok(bytes) => {
+                if bytes == 0 {
+                    Poll::Ready(None)
+                } else {
+                    Poll::Ready(Some(Ok(hyper::body::Bytes::from(buf[0..bytes].to_vec()))))
+                }
+            }
+            Err(e) => Poll::Ready(Some(Err(e))),
+        }
+    }
+}
+
+/// Send a stream of file to client.
+pub fn send_file<P: AsRef<Path>>(file_path: P) -> io::Result<(FileStream<BufReader<File>>, u64)> {
+    let file = File::open(file_path)?;
+    let size = file.metadata()?.len();
+    let reader = Mutex::new(BufReader::new(file));
+    Ok((FileStream { reader }, size))
 }
 
 /// Sending a directory as zip buffer
 pub fn send_dir_as_zip<P: AsRef<Path>>(
     dir_path: P,
     show_all: bool,
     with_ignore: bool,
-) -> io::Result<Vec<u8>> {
+) -> io::Result<(FileStream<BufReader<File>>, u64)> {
     let dir_path = dir_path.as_ref();
 
     // Creating a memory buffer to make zip file
-    let mut zip_buffer = Vec::new();
-    let cursor = std::io::Cursor::new(&mut zip_buffer);
+    let zip_file = tempfile::tempfile()?;
```

Comment:
Nice trick!

Can you also update the comment above this line?

Reply:
updated on 481b1fb